### PR TITLE
MSC2732: Olm fallback keys

### PR DESCRIPTION
**Pull request by uhoreg**
[Rendered](https://github.com/uhoreg/matrix-doc/blob/fallback_keys/proposals/2732-olm-fallback-keys.md)

[FCP proposal](https://github.com/matrix-org/matrix-doc/pull/2732#issuecomment-705053777)

### CODE REVIEW

File: proposals/2732-olm-fallback-keys.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+given algorithm, it is replaced -- the server will only keep one fallback key
+per algorithm for each user.
+
+When Bob calls `/keys/claim` to claim one of Alice's one-time keys, but Alice
+has no one-time keys left, the homeserver will return the fallback key instead,
+if Alice had previously uploaded one.  Unlike with one-time keys, fallback keys
+are not deleted when they are returned by `/keys/claim`.  However, the server
+marks that they have been used.
+
+A new response parameter, `device_unused_fallback_keys`, is added to `/sync`.
```
Comment (auscompgeek) [inline proposals/2732-olm-fallback-keys.md:27]: How should the absence of this field be interpreted?

Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:27]: I wasn't sure if I should clarify it in the text, but maybe I should.  Any server that complies with whatever spec version this makes it into must include the parameter (and it should be an empty array if there are no unused fallbacks).  If the server doesn't include the key, then the client can assume that the server doesn't comply with that version of the spec -- in particular, it can assume that the server doesn't support fallback keys, and so should not try to upload fallbacks to the server (or know that the server will ignore any fallbacks that it did upload).

**Comment by uhoreg** [comment]
Implemented by https://gitlab.matrix.org/matrix-org/olm/-/merge_requests/13, https://github.com/matrix-org/matrix-js-sdk/pull/1467, and https://github.com/matrix-org/synapse/pull/8312

File: proposals/2732-olm-fallback-keys.md (hunk 1 above)
Comment (jryans) [inline proposals/2732-olm-fallback-keys.md:27]: Does this mean a supporting server will list your fallback key in every sync response indefinitely? It would be nice if there were some way for it to only appear when changed to avoid adding overhead to every sync response.

Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:27]: > Does this mean a supporting server will list your fallback key in every sync response indefinitely?

Yes (until it's used).  This makes it easier for client authors to keep the fallback keys up to date, as they just need to check if the array contains the key type expected, and if not, the client uploads a new key.  This works for both initial setup (so the client doesn't need to remember if it already uploaded a key), and ensuring that the server has a new key.

This also mirrors the existing `device_one_time_keys_count` property.

File: proposals/2732-olm-fallback-keys.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+`one_time_keys` parameter with the exception that there must be at most one key
+per key algorithm.  If the user had previously uploaded a fallback key for a
+given algorithm, it is replaced -- the server will only keep one fallback key
+per algorithm for each user.
+
+When Bob calls `/keys/claim` to claim one of Alice's one-time keys, but Alice
+has no one-time keys left, the homeserver will return the fallback key instead,
+if Alice had previously uploaded one.  Unlike with one-time keys, fallback keys
+are not deleted when they are returned by `/keys/claim`.  However, the server
+marks that they have been used.
```
Comment (dbkr) [inline proposals/2732-olm-fallback-keys.md:25]: Would there be any value to Bon knowing that he has received a fallback key? eg. his client could warn him that if he has anything *really* secret to tell Alice, he might want to wait?

File: proposals/2732-olm-fallback-keys.md (hunk 1 above)
Comment (dbkr) [inline proposals/2732-olm-fallback-keys.md:27]: It seems fairly trivial but an example could still be nice.

File: proposals/2732-olm-fallback-keys.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+A new response parameter, `device_unused_fallback_keys`, is added to `/sync`.
+This is an array listing the key algorithms for which the server has an unused
+fallback key for the user.  If the client wants the server to have a fallback
+key for a given key algorithm, but that algorithm is not listed in
+`device_unused_fallback_keys`, the client will upload a new key as above.
```
Comment (dbkr) [inline proposals/2732-olm-fallback-keys.md:31]: Do we need a way to remove the fallback key of the client changes it's mind about this?

File: proposals/2732-olm-fallback-keys.md (hunk 2 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:25]: Possibly.  One way to do that would be to have the client that uploads the fallback key to add a field to the key (assuming the key is in the form of a JSON object) saying that it's a fallback key (e.g. `"fallback": true`)  In this way, the server cannot lie about the key being a fallback key or not, since it's covered by the key signature.

File: proposals/2732-olm-fallback-keys.md (hunk 3 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:55]: I'm not sure if it's necessary, but should be fairly easy to add later on if we want, so I think we can just leave it off for now.

File: proposals/2732-olm-fallback-keys.md (hunk 2 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:49]: I've added a thing to the doc saying this.

File: proposals/2732-olm-fallback-keys.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+```
+
+When Bob calls `/keys/claim` to claim one of Alice's one-time keys, but Alice
+has no one-time keys left, the homeserver will return the fallback key instead,
+if Alice had previously uploaded one.  Unlike with one-time keys, fallback keys
+are not deleted when they are returned by `/keys/claim`.  However, the server
+marks that they have been used.
+
+A new response parameter, `device_unused_fallback_keys`, is added to `/sync`.
```
Comment (clokep) [inline proposals/2732-olm-fallback-keys.md:51]: It was mentioned in https://github.com/matrix-org/synapse/pull/8312#discussion_r498492265 that this is really the *types* of those keys, not the full keys and suggested to change this to `device_unused_fallback_key_types`. This doesn't yet seem reflected in the MSC (although @uhoreg said it was updated, so maybe it just never got pushed).

Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md]: Indeed, I forgot to push

**Comment by uhoreg** [comment]
@mscbot fcp merge

File: proposals/2732-olm-fallback-keys.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# MSC2732: Olm fallback keys
```
Comment (joepie91) [inline proposals/2732-olm-fallback-keys.md:1]: It's not quite clear to me what exactly the purpose and tradeoffs of this entire design are (hence attaching this thread to the first line, as good a place as any).

What does this improve over just having a number of one-time keys? As I understand it, the benefit is that fallback keys can be used indefinitely and so there is no limit on established sessions like with pre-generated keys.

But then why not drop the pre-generated key mechanism in favour of this mechanism entirely? As I understand it, because the security guarantees of this model are weaker, and so it is preferable to use pre-generated keys where possible.

But then doesn't this weaken the overall security model, by making it possible for an attacker to intentionally exhaust all of someone's pre-generated keys, essentially carrying out a downgrade attack and forcing them (or rather, the people communicating with them) into fallback keys being used instead, which would weaken security properties?

I have difficulty squaring this circle and understanding how this can be both a) useful, b) secure, *and* c) additive to the current model. Can you shed some light on this?

File: proposals/2732-olm-fallback-keys.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+name, a one-time key is only to be used once.  However, if all of Alice's
+one-time keys are claimed, Bob will not be able to create a session with Alice.
+
+This can be addressed by Alice uploading a fallback key that is used in place
+of a one-time key when no one-time keys are available.
+
+## Proposal
+
+A new request parameter, `fallback_keys`, is added to the body of the
+`/keys/upload` client-server API, which is in the same format as the
```
Comment (richvdh) [inline proposals/2732-olm-fallback-keys.md:15]:
```suggestion
[`/keys/upload` client-server API](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-keys-upload), which is in the same format as the
```

File: proposals/2732-olm-fallback-keys.md (hunk 5 above)
Comment (richvdh) [reply proposals/2732-olm-fallback-keys.md:1]: +1. I have vague memories from staring at the signal protocol (which does the same trick) that this wasn't as much of a disaster as you might think, but I can't remember why. @uhoreg I think this is worth spelling out.

Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:1]: Right, I keep forgetting to respond to this.

> But then why not drop the pre-generated key mechanism in favour of this mechanism entirely? As I understand it, because the security guarantees of this model are weaker, and so it is preferable to use pre-generated keys where possible.

My understanding is that it is slightly weaker security, but it's not a huge difference.  So it is "better" to use one-time keys, but when one is unavailable, a fallback key is "good enough" in many cases.  But this proposal allows clients to make up their own mind about the tradeoffs.  If they don't think that a fallback key is secure enough, then they don't need to use it.  If they don't care at all about the extra security from one-time keys vs. fallback keys, they can drop one-time keys completely and just rely on fallback keys.  If they want to the extra security from one-time keys when available, but don't want to inconvenience the user when one-time keys aren't available, it can choose to do that too.

As an explanation of the security difference: *under an assumption that an attacker cannot break Curve25519*, and so must attack the client directly to get the private keys, there is not much difference between using a one-time key and a fallback key.  If an attacker is able to extract the keys from the client, then they will have the ciient's private identity key, all the one-time keys that have already been used, and the current fallback key.  And the only difference between having a private key for a one-time key and for a fallback key is that the fallback key may have been used already in a session that was already processed.  But if a client promptly replaces a fallback key after it has been used and forgets the private key quickly (after it's reasonably sure that it has received all the sessions that use it), then the difference is small.

So I think this comes down to a case of cryptographic ideal vs. pragmatism.  From a cryptographic standpoint, one-time keys are the way to go, because we're paranoid and worry about Curve25519 being broken.  But practically speaking, the tradeoff between the possibility of someone cracking Curve25519 or attacking the client at just the right moment that they can decrypt some extra sessions, versus the inconvenience of having undecryptable messages because we ran out of one-time keys, for most people, I think, leans towards convenience in this case.  But again, the client can make they choice about what to do, and the user can choose a client that matches their paranoia level.

File: proposals/2732-olm-fallback-keys.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+the previous fallback key, and should remove the previous fallback key once it
+is reasonably certain that it has received all the messages that use it (for
+example, one hour after receiving the first message that used it).
+
+For addressing replay attacks, clients can also keep track of inbound sessions
+to detect replays.
+
+## Unstable prefix
+
+The `fallback_key` request parameter and the `device_unused_fallback_key_types`
```
Comment (Sorunome) [inline proposals/2732-olm-fallback-keys.md:96]:
```suggestion
The `fallback_keys` request parameter and the `device_unused_fallback_key_types`
```

File: proposals/2732-olm-fallback-keys.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+name, a one-time key is only to be used once.  However, if all of Alice's
+one-time keys are claimed, Bob will not be able to create a session with Alice.
+
+This can be addressed by Alice uploading a fallback key that is used in place
+of a one-time key when no one-time keys are available.
+
+## Proposal
+
+A new request parameter, `fallback_keys`, is added to the body of the
+[`/keys/upload` client-server API](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-keys-upload), which is in the same format as the
```
Comment (Sorunome) [inline proposals/2732-olm-fallback-keys.md:15]: Does this also add any new response to having successfully uploaded the fallback keys, similar to OTKs?

File: proposals/2732-olm-fallback-keys.md (hunk 5 above)
Comment (Sorunome) [reply proposals/2732-olm-fallback-keys.md:1]: with OTKs you have to break every single curve25519 keypair. With fallback keys if you break the curve25519 fallback key you immidiately have all sessions with devices who used that fallback key to initiate communication.

Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:1]: Yeah.  If the attacker is able to break the Curve25519 key with near-zero effort, then having OTKs doesn't really help at all.  But the difference comes when the attacker needs to spend some time/effort to break it, where the time is somewhere between "so short that it's trivial" and "too long to be practical".  Then having OTKs means that they need to spend that time for each keypair that they want to break.

Comment (erikjohnston) [reply proposals/2732-olm-fallback-keys.md:1]: Interesting, some of that is going over my head. It does somewhat sound like OTKs vs fallback keys provide the same security in the case where you rotate the fallback key relatively quickly after someone uses it?

Comment (erikjohnston) [reply proposals/2732-olm-fallback-keys.md:1]: So the only time the difference is important is if multiple new sessions come along (but less than the number of OTKs you have uploaded) before you open up your client again. This feels like quite a small difference in security that I'm leaning towards it being pointless to have both?

**Comment by erikjohnston** [comment]
@mscbot concern Do we need both OTKs if we have fallback keys? ─ https://github.com/matrix-org/matrix-doc/pull/2732#discussion_r519022271

File: proposals/2732-olm-fallback-keys.md (hunk 5 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:1]: I don't know if I'm ready to boot out OTKs right away, and I think that it would be OK to leave it up to client developers to decide how paranoid they want to be for now.  We can remove OTKs in a future MSC if we need to.

Comment (erikjohnston) [reply proposals/2732-olm-fallback-keys.md:1]: I think its fair to take things slow, and we can't throw away OTKs immediately due to backwards compat. I'm just very cautious about the spec accumulating multiple different ways of doing the same thing, *especially* with the crypto code. (I can see situations where one client decides "for security" to never user fallback keys, and another "for simplicity" to never use OTKs, and now they can't talk).

Comment (poljar) [reply proposals/2732-olm-fallback-keys.md:1]: Removing OTKs doesn't seem sensible to me, the necessity for backwards compat makes this not worth it even if there are no downsides otherwise. You'll have to support both methods to calculate the shared secret at least for a while.

Let's reiterate what one-time keys buy us, from the [Signal x3dh spec](https://signal.org/docs/specifications/x3dh/#key-compromise):

>If one-time prekeys are used for a protocol run then a compromise of Bob's identity key and prekey private keys at some future time will not compromise the older SK, assuming the private key for OPKB was deleted.

Vs what happens if one-time keys were not used:

> If one-time prekeys were not used for a protocol run, then a compromise of the private keys for IKB and SPKB from that protocol run would compromise the SK that was calculated earlier. Frequent replacement of signed prekeys mitigates this, as does using a post-X3DH ratcheting protocol which rapidly replaces SK with new keys to provide fresh forward secrecy [5].

So looking at this, why would we want to have weaker forward secrecy? Sure one could exhaust one-time keys when you do want to create a shared secret and force a downgrade that way. It's important to understand that forward secrecy aims to protect older shared secrets from future compromises, so the assumption is that the attack and possible OTK exhaustion will happen in the future.

To be honest I would prefer Olm to stick to well understood, and well researched ways to establish the shared secret that Signal uses. Put it in other words, if we're going to change how the shared secret should be calculated, use the fallback key and one-time keys, or use just the fallback key, instead of using just one or the other.

We have a couple of differences from Signal already and I can't really find a lot of rationale for why things were changed up. The removal of the fallback key is such a change which is now giving us headaches. Another such example is this line from the [Olm spec](https://gitlab.matrix.org/matrix-org/olm/-/blob/master/docs/olm.md#version-1):

> The length of the MAC is determined by the authenticated encryption algorithm being used. (Olm version 1 uses HMAC-SHA-256, truncated to 8 bytes). The MAC protects all of the bytes preceding the MAC.

While the [Signal spec](https://signal.org/docs/specifications/doubleratchet/#truncating-authentication-tags) has this line:

> If the ENCRYPT() function is implemented using CBC and HMAC as described in Section 5.2, then truncating the final HMAC output to 128 bits to reduce message size is acceptable. Truncating it further might be acceptable, though requires careful analysis. In no case should the final HMAC be truncated to less than 64 bits.

Please correct me if I'm wrong, but I'm not sure we can point people to such careful analysis, people did ask about this a couple of times. I feel like removing OTKs would be a similar situation like this one.

Comment (erikjohnston) [reply proposals/2732-olm-fallback-keys.md:1]: > Removing OTKs doesn't seem sensible to me, the necessity for backwards compat makes this not worth it even if there are no downsides otherwise. You'll have to support both methods to calculate the shared secret at least for a while.

I think we need to be really careful about accumulating cruft over time, its effectively tech debt in spec form. Yes, we'd need a (fairly long) transition period, but we also need to be able to make changes to these protocols.


> So looking at this, why would we want to have weaker forward secrecy? Sure one could exhaust one-time keys when you do want to create a shared secret and force a downgrade that way. It's important to understand that forward secrecy aims to protect older shared secrets from future compromises, so the assumption is that the attack and possible OTK exhaustion will happen in the future.

I don't really understand what you mean. Does only using the fallback key *actually* protect against any attacks? Especially if an attacker can force pool exhaustion. Having more moving parts introduces its own risk, especially with multiple independent implementations in the wild. Given that the fallback key should live for the same amount of time as the OTKs (i.e. it gets rotated once the app comes online and sees that its been used), I find it hard to see how someone would compromise a fallback key without also compromising all current OTKs. Indeed, fallback keys don't need to be precomputed, so a compromise would only give the *current* fallback key that may only be used once, rather than *all* current OTKs.

The only benefit to OTKs that their design effectively forces them to be correctly rotated, whereas clients could fail to correctly rotate their fallback keys.

I am sympathetic to the fact that we shouldn't be diverging from the Signal spec though. We certainly shouldn't be diverging without strong justification.

Comment (Sorunome) [reply proposals/2732-olm-fallback-keys.md:1]: > (I can see situations where one client decides "for security" to never user fallback keys, and another "for simplicity" to never use OTKs, and now they can't talk).

They would still be able to talk to each other. OTKs and fallback keys only differentiate when uploading / rotating. For a client *using* them, they look identical.

Comment (erikjohnston) [reply proposals/2732-olm-fallback-keys.md:1]: > > (I can see situations where one client decides "for security" to never user fallback keys, and another "for simplicity" to never use OTKs, and now they can't talk).
> 
> They would still be able to talk to each other. OTKs and fallback keys only differentiate when uploading / rotating. For a client _using_ them, they look identical.

Aha, missed that point, thanks!

Comment (richvdh) [reply proposals/2732-olm-fallback-keys.md:1]: > We have a couple of differences from Signal already and I can't really find a lot of rationale for why things were changed up. The removal of the fallback key is such a change which is now giving us headaches. 

I want to correct a misconception here: Signal added fallback keys to their protocol *after* we forked it into Olm. The same might apply to the MAC length recommendation; I don't remember offhand.

So, the relevant question here isn't "why were fallback keys stripped out", but rather "why didn't we immediately follow suit when Signal added them", which can be answered as (a) lack of tuits, but more fundamentally (b) I don't accept that we should automatically do everything Signal does just because Signal does it. They have some excellent ideas cryptographically, but at the end of the day their product is a bit different from ours, and certainly some of their decisions have been at least open to debate in the past.

Anyway, this is a bit off-topic. It's certainly correct that we should look at Signal's use of fallback keys and consider whether we should follow suit or diverge, and we should probably have done it sooner. That is part of what this MSC is about, so I now return you to your regular programming.

Comment (poljar) [reply proposals/2732-olm-fallback-keys.md:1]: Sorry to keep on dwelling on this, but I think we'll all spend less time on this if I can squash the potential for an MSC that intends to remove one-time keys right here.

> I don't really understand what you mean. Does only using the fallback key _actually_ protect against any attacks?

Did you mean "only using one-time keys"? It does not protect against any novel attack that the rotation of fallback keys doesn't protect against, if you rotate them. It just does so better.

> Especially if an attacker can force pool exhaustion. Having more moving parts introduces its own risk, especially with multiple independent implementations in the wild. Given that the fallback key should live for the same amount of time as the OTKs (i.e. it gets rotated once the app comes online and sees that its been used), I find it hard to see how someone would compromise a fallback key without also compromising all current OTKs. Indeed, fallback keys don't need to be precomputed, so a compromise would only give the _current_ fallback key that may only be used once, rather than _all_ current OTKs.

That's not what forward secrecy is about, the term is a bit confusing and it trips me often as well, but this is a nice short reminder what we're talking about:

> Forward secrecy protects past sessions against future compromises of keys or passwords.

So given that we have established shared secrets, a compromise of secret keys in the future wouldn't compromise the shared secret. In this case the one-time keys that established the session are gone, if the fallback key is used it might not be gone.

> The only benefit to OTKs that their design effectively forces them to be correctly rotated, whereas clients could fail to correctly rotate their fallback keys.

I don't think this is correct and I'm not sure if I'm bad at explaining things or my English doesn't work as well as I think it does, but let's look at the implementation, perhaps that's going to clear up things.

One time keys are useful because they can be used only once, now bear with me for a while and don't throw your computer through the window just yet. In other words, this means that only a single shared secret should be created with a given one-time key. If we use the key, we can throw it away.

In contrast, fallback keys can be used many times. We don't have a way of knowing how many shared secrets will be established using the fallback key. Given that we don't know how many people have used our fallback key it becomes hard to know when we can throw the key away. Let's say we rotate the key every time we notice that it has been used, we still can't just throw away the old key since there's going to be a period of time where we upload a fresh key but people may still use the old one.

This in turn means we'll have at least two fallback keys around in the general case (please note that this is what has been implemented in libolm in [this](https://gitlab.matrix.org/matrix-org/olm/-/commit/171044f3fca084bd1c5c1f8f4bfe146434b07f0b#76a7620a7a14f3dc912f2778aaf27201b55421a4_46_46) commit. I would argue that, if one-time keys are removed, you'll have to keep even more fallback keys around, given our decentralized nature.

This leads us to the fact that no matter how quickly you rotate your fallback keys, they won't be gone as fast as one-time keys. Thus one-time keys provide stronger forward secrecy than just using fallback keys that you rotate.

I'll stop being off topic in this MSC now and won't answer here anymore. If you have further questions please forward them to me in the E2E Matrix room, or over email. Alternatively if you still think that this is a good idea an MSC to remove the one-time keys can be the next discussion place for this.

Comment (poljar) [reply proposals/2732-olm-fallback-keys.md:1]: > I want to correct a misconception here: Signal added fallback keys to their protocol _after_ we forked it into Olm. The same might apply to the MAC length recommendation; I don't remember offhand.

Is this correction of the misconception correct (🥁)? I have not been around back then but as far as I can tell libolm was forked from Axolotl (v2?) which did have a concept of a fallback key, Axolotl (v3?) which introduced x3dh
appeared a couple of months after the fork. The removal of the fallback key can be found in [this](https://gitlab.matrix.org/matrix-org/olm/-/commit/f1cd5613f68beee08ab479edf2a5ae50c91a63fe) commit.

I don't know if the truncation has been changed or not, but I think that one case of such changes (the removal of the fallback key) and one potential such change (the removal of one-time keys) is enough to get my point across.

> So, the relevant question here isn't "why were fallback keys stripped out", but rather "why didn't we immediately follow suit when Signal added them", which can be answered as (a) lack of tuits, but more fundamentally (b) I don't accept that we should automatically do everything Signal does just because Signal does it. They have some excellent ideas cryptographically, but at the end of the day their product is a bit different from ours, and certainly some of their decisions have been at least open to debate in the past.

I don't think this reversal is true given the above so I won't address everything, but my complaint isn't about following Signal step by step, it's about introducing changes to complex protocols without proper justification.

Fallback keys are used to defend against denial of service attacks. We should know why we don't want this property in the protocol.

One-time keys are used to offer better forward secrecy. Again, we should know why this isn't desirable.

Comment (erikjohnston) [reply proposals/2732-olm-fallback-keys.md:1]: > > Forward secrecy protects past sessions against future compromises of keys or passwords.
> 
> So given that we have established shared secrets, a compromise of secret keys in the future wouldn't compromise the shared secret. In this case the one-time keys that established the session are gone, if the fallback key is used it might not be gone.

To be almost unhelpfully pedantic here, OTKs *don't* strictly allow this as they're not deleted immediately on key use, they're deleted once the client has become aware that they've been used. Sure, this is likely a short window of time, but then if the fallback keys are only kept for a short amount of time the situations become effectively equivalent. I don't believe any of our forward secrecy actually immediately protects *all* old messages, rather it protects sufficiently old messages, i.e. there's always a period of time in the past where an attacker can still get messages for. 

This is the crux of what confuses me: if OTKs only slightly reduce this window of time, and we're saying we're happy with the window that having fallback keys gives us (since we're using them), then are OTKs worth the additional complexity?

However:

> Given that we don't know how many people have used our fallback key it becomes hard to know when we can throw the key away.

I think is what I've been missing here, I've been assuming that we can delete the fallback private keys quickly, and so the window of attack for OTKs vs fallback keys are effectively the same. If we have to keep round the fallback keys for hours, then that's probably(?) a sufficiently large window that using OTKs to reduce that windows makes sense.

The other piece here is that while an attacker can drain a device's OTKs, and so force new sessions to use the fallback keys, that is an *active* attack that can be observed by the servers and device. That will at least give some breadcrumbs that something fishy is going on, rather than allowing completely passive attacks. Though since draining of OTKs happens sufficiently often that we're making this MSC, I don't know if anybody would actually notice the draining of OTKs.

Basically: adding fallback keys weakens security, and while removing OTKs would weaken security some more, do they provide enough meaningful protection to warrant their complexity? It sounds like the answer is "yes, just", so I'm OK keeping them long term if people agree with the above analysis.

---

> I'll stop being off topic in this MSC now and won't answer here anymore.

I'd rather these thoughts were recorded on the MSC for future reference, so that they can be linked back.

Comment (richvdh) [reply proposals/2732-olm-fallback-keys.md:1]: > The removal of the fallback key can be found in this commit.

oh. I'll get back in my box. Sorry.

Comment (dkasak) [reply proposals/2732-olm-fallback-keys.md:1]: > I think is what I've been missing here, I've been assuming that we can delete the fallback private keys quickly, and so the window of attack for OTKs vs fallback keys are effectively the same. If we have to keep round the fallback keys for hours, then that's probably(?) a sufficiently large window that using OTKs to reduce that windows makes sense.

Worse than that, since there is no way to determine the number of times the fallback key was retrieved (as far as I'm aware), we have no deterministic way of concluding that all sessions involving that fallback key have been created. So the choices are between keeping a given fallback key for an indefinite amount of time or else risk having undecryptable messages.

This doesn't happen with OTKs due to their property that they will be used at most once, so when a session is created for a given OTK, the client can conclude with certainty that it can drop that OTK.

**Comment by erikjohnston** [comment]
@mscbot resolve Do we need both OTKs if we have fallback keys? ─ https://github.com/matrix-org/matrix-doc/pull/2732#discussion_r519022271

File: proposals/2732-olm-fallback-keys.md (hunk 5 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:1]: > Worse than that, since there is no way to determine the number of times the fallback key was retrieved (as far as I'm aware), we have no deterministic way of concluding that all sessions involving that fallback key have been created. So the choices are between keeping a given fallback key for an indefinite amount of time or else risk having undecryptable messages.

Generally, it should be fairly safe to assume that when a user claims a one-time key (whether it's actually an OTK or ends up being a fallback key), they're going to use it right away.  So clients should only need to keep the fallback key for a little time after it's used (maybe a couple minutes after the sync in which they notice that it's been used, to allow for network delays).  The risk of having undecryptable messages is somewhat mitigated by having olm unwedging and key resharing, though that requires the sender to come back online.
 
> This doesn't happen with OTKs due to their property that they will be used at most once, so when a session is created for a given OTK, the client can conclude with certainty that it can drop that OTK.

We actually do have a similar problem with OTKs.  libolm tries to use a constant amount of memory, so it only has limited space for OTKs.  That means that libolm will sometimes evict OTKs when it generates new ones, so if someone claims a OTK and waits too long to use it, it may have been evicted by the time they finally get around to using it.

File: proposals/2732-olm-fallback-keys.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+When Bob calls `/keys/claim` to claim one of Alice's one-time keys, but Alice
+has no one-time keys left, the homeserver will return the fallback key instead,
+if Alice had previously uploaded one.  Unlike with one-time keys, fallback keys
+are not deleted when they are returned by `/keys/claim`.  However, the server
+marks that they have been used.
+
+A new response parameter, `device_unused_fallback_key_types`, is added to
+`/sync`.  This is an array listing the key algorithms for which the server has
+an unused fallback key for the user.  If the client wants the server to have a
```
Comment (richvdh) [inline proposals/2732-olm-fallback-keys.md:53]: forgive what might be a stupid question: what constitutes an `unused` fallback key? One that has not yet been given out over `/keys/claim`?

File: proposals/2732-olm-fallback-keys.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+that the server supports fallback keys, and so only upload fallback keys to
+servers that support them.
+
+Example:
+
+`GET /sync`
+
+Response:
+
+```json
```
Comment (richvdh) [inline proposals/2732-olm-fallback-keys.md:68]: This will make github happy about the comment:

```suggestion
```jsonc
```

File: proposals/2732-olm-fallback-keys.md (hunk 9 above)
Comment (richvdh) [inline proposals/2732-olm-fallback-keys.md:53]:
```suggestion
an unused fallback key for the device.  If the client wants the server to have a
```
I assume?

File: proposals/2732-olm-fallback-keys.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+still active.  Also, an attacker that compromises a client may be able to
+retrieve the private part of the fallback key to decrypt past messages if the
+client has still retained the private part of the fallback key.
+
+For this reason, clients should not store the private part of the fallback key
+indefinitely.  For example, client should only store at most two fallback keys:
+the current fallback key (that it has not yet received any messages for) and
+the previous fallback key, and should remove the previous fallback key once it
+is reasonably certain that it has received all the messages that use it (for
+example, one hour after receiving the first message that used it).
```
Comment (richvdh) [inline proposals/2732-olm-fallback-keys.md:89]: is the client supposed to rotate the fallback key as soon as the current one is used? If so, that implies it needs to store 
if you have to keep the previous fallback key for an hour, and you're only allowed to store two keys, that implies you should only rotate them once an hour at most, which is in conflict with "should only store at most two fallback keys".

File: proposals/2732-olm-fallback-keys.md (hunk 9 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:53]: Yes.  I can clarify that.

File: proposals/2732-olm-fallback-keys.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
```
Comment (richvdh) [inline proposals/2732-olm-fallback-keys.md:89]: I've no objections to this from the POV of the protocol, but still think it would be nice to clarify how key rotation is expected to work.

**Review by richvdh — APPROVED** [review]

**Review by erikjohnston — APPROVED** [review]

File: proposals/2732-olm-fallback-keys.md (hunk 11 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:89]: > is the client supposed to rotate the fallback key as soon as the current one is used?

Yes.  We could say that it should be rotated after it has done processing the `/sync` request in which the current fallback key was used.  At the same time, it should replenish the one-time keys.  So my thinking was that it would be unlikely (under normal circumstances) that all the one-time keys get exhausted in an hour.

However, this changes if the client opts not to use one-time keys and to only use fallback keys.  And an attacker could possibly run through all the one-time keys quickly.  So, I think we could either

- reduce the time recommended for a client to keep the fallback key, maybe to ~5 minutes, and rate-limit rotating the key.  If all new sessions created within a 5-minute period end up using the same key, that doesn't seem to terrible.
- say that if you receive a new session created using the new fallback key, then discard the old fallback key, regardless of when it was last rotated -- if a session comes in with the new fallback key, it's probably less likely that there will be remaining sessions using the old fallback key, unless there are major federation delays (clients should generally start a new session very shortly after claiming the key, so the delay between the user receiving the session should be fairly short).

I think that I prefer the first option.  I can't really think of other options, aside from variations of the theme, that allow an upper bound on the number of keys a client has to keep.

File: proposals/2732-olm-fallback-keys.md (hunk 8 above)
Comment (uhoreg) [reply proposals/2732-olm-fallback-keys.md:15]: We don't currently have any addition to the response.

**Comment by turt2live** [comment]
Spec PR: https://github.com/matrix-org/matrix-doc/pull/3615

**Comment by turt2live** [comment]
Merged 🎉
